ConvertCiv3Media/ReadPcx.cs

```diff
@@ -43,44 +43,48 @@ public byte ColorIndexAt(int x, int y)
             int PaletteOffset = PcxBytes.Length - 768;
 
             // Populate color palette
-            for (int i = 0; i < 256; i++) {
-                this.Palette[i,0] = PcxBytes[PaletteOffset + i * 3];
-                this.Palette[i,1] = PcxBytes[PaletteOffset + i * 3 + 1];
-                this.Palette[i,2] = PcxBytes[PaletteOffset + i * 3 + 2];
-            }
+            Buffer.BlockCopy(PcxBytes, PaletteOffset, Palette, 0, 768);
+
+            int Length = Width * Height;
+            ColorIndices = new byte[Length];
 
-            // Populate image byte array
-            List<byte> ListImage = new List<byte>();
             // Encoding always have even number of bytes per line; if image width is odd, there is a junk byte in every row
             bool JunkByte = BytesPerLine > Width;
+            byte PcxData;
+
             // Loop to decode run-length-encoded image data which begins at file offset 0x80
-            for (int ImgIdx = 0, PcxIdx = 0x80, RunLen = 0, LineIdx = 0; ImgIdx < Width * Height; ) {
+            for (int ImgIdx = 0, PcxIdx = 0x80, RunLen = 0, LineIdx = 0; ImgIdx < Length; ) {
                 // if two most significant bits are 11
-                if ((PcxBytes[PcxIdx] & 0xc0) == 0xc0) {
+                PcxData = PcxBytes[PcxIdx];
+                if ((PcxData & 0xc0) == 0xc0) {
                     // then it & 0x3f is the run length of the following byte
-                    RunLen = PcxBytes[PcxIdx] & 0x3f;
+                    RunLen = PcxData & 0x3f;
                     PcxIdx++;
                     // Repeat the pixel in the image RunLen times
                     for (int j = 0; j < RunLen; j++) {
                         // Add pixel copy if it's not a junk byte
-                        if (!(JunkByte && LineIdx % BytesPerLine == BytesPerLine - 1)) {
-                            ListImage.Add(PcxBytes[PcxIdx]);
-                            ImgIdx++;
+                        if (JunkByte) {
+                            if (++LineIdx == BytesPerLine) {
+                                LineIdx = 0;
+                                ImgIdx--;
+                            }
```

Comment:
I'm not really sure if the new code is any clearer than the old code, which admittedly was not very clear.  So I took a look at what I wrote in Java back in 2012 to get an idea of what might be clearer.  A few thoughts:

 - JunkByte is confusing.  I initially thought it was "is this byte a junk byte?", but it's just a true/false of whether there are junk bytes.  Suggestions below, based on the other thoughts...
 - Prefix/Postfix operators (++x, x++), when combined with an operation on the same line, are IMO confusing.  I always have to check "does it add first when it's ++x or x++?"  I'm guessing it must be adding first before comparing to BytesPerLine here, but not having to think about that would simplify the relatively complex code a bit.  I'd rather have an extra line than have to think about that in other words.  (If I were primarily a low-level programmer this probably wouldn't be of concern, but IIRC only Flintlock is really a low-level programmer on the team)
 - In the new code, we're still adding the byte to ColorIndices if it's a junk byte; in the old code we are not.  I'm 99% sure the old code is correct in that regard.
 - My Java code appears to not handle junk bytes at all.  Why doesn't it fail?  Probably because all the Civ3 graphics have an even width.  This would also be why the above didn't have any visual side effects.

So let's see if I can re-write this following my suggestions, while still having it be nicer than the old version.

```
for (int j = 0; j < RunLen; j++) {
    if (!(JunkByte && (LineIdx == BytesPerLine - 1))) {
        ColorIndices[ImgIdx] = PcxBytes[PcxIdx];
        ImgIdx++;
    }
    LineIdx++;
    if (LineIdx == BytesPerLine) {
        LineIdx = 0;
    }
}
```
Hmm... that wound up being almost the same as the old code.  So maybe the suggestion should be "add a helper function to make things more clear", e.g.

```
for (int j = 0; j < RunLen; j++) {
    if (shouldAddPixel(JunkByte, LineIdx, BytesPerLine)) {
        ...

bool shouldAddPixel(bool JunkByte, int LineIdx, Int BytesPerLine) {
    return !(JunkBytes && (LineIdx == BytesPerLine - 1));
}
```
That could then be used in the literal case as well.

Also maybe DiscardLastByte instead of JunkByte?  It's bothering me that the current name implies the current byte is a junk byte when that isn't the case.

Reply:
Haha, I see now why my Java code doesn't worry about junk bytes.  It uses 2D indexing, and simply reads in the junk bytes into its array.  But then when it goes to convert that 2D array of indexes to an image, it only reads the x indices through xSize, rather than through BytesPerLine.

That wouldn't work with our 1D array of indices, and I suspect when I read the other file I'll see that the 1D aspect is important for performance.  But it's an interesting tradeoff of where the complexity lies.

Reply:
Ah, that is a fun way to circumvent the junk byte problem! But yeah, they do matter. Commenting out that part of the code causes the program to crash when trying to load a sav (though the main menu works fine), so there are at least some odd pcx files.

I admit this isn't any clearer than the old code and is probably worse in fact to read. I went a little bit crazy with the micro-optimizations here, since reverting all changes to this main PCX loop (except the switching from List to array) only slows the "Game scene load time" down by about 40-50ms. Now, this isn't a full measure of the slowdown, since PCX files are loaded and will be loaded in other places, but even then, it's not a crazy difference. Personally, my style-vs-performance tradeoff calculus still favors keeping the changes, because having measurable differences left on the table eats away at me, but I also completely get why that's not necessarily the right call for a long-term project.

Reply:
As far as working correctly, the new code is actually producing the same correct results as the old. It's just doing it in a very sneaky and not-at-all-sufficiently-documented way. Like you say, when the loop goes into single-pixel mode instead of run-of-pixels mode, it sets the next ColorIndices value, regardless of whether it's a junk byte or not. But then it *does* check whether that byte was a junk value, and if so, it decrements the ImgIdx. What this means is that on the next loop iteration, the same pixel will be written to, correcting post-hoc any mistakes. The reason why the ColorIndices doesn't overflow when the end of it is reached and there is still an extra junk byte left to go is that the outer loop stops when width * height number of pixels have been read. So yeah, this is an equivalent change to the code, but perhaps a poorly-designed one.

Reply:
Oh, I see now.  It decrements ImgIndex, so the next (non-junk) byte will overwrite the junk byte.  So yes, it will work (and I really ought to manually verify that, too...).

Yeah, I'd say if we get 95-96% of the savings without the micro-optimizations, it isn't worth keeping the ones that make it harder to read, even in a slow way.  Some of them don't affect readability, so we'd probably wind up with 98% or so of the improvement.  I try to keep in mind Kernighan's quip about debugging being more difficult than writing code, and being as clever as possible when writing it means you can't debug it, and also remembering that whoever comes later may not be as clever with regards to whichever areas we have expertise in.

I also try to apply Amdahl's Law, which generally points me to macro-optimizations.  E.g. I've been loading saves from The Great War locally, and it takes about 8 seconds.  I haven't broken down why it takes so long, but I'm pretty sure most of the different isn't in PCX loading.  Once this is merged, we can apply some high level metrics and see what's next in the order of making loads take awhile... although we probably should get around to the SAV pull request before that point since that may have an impact as well.